File: 141-linked-list-cycle/141-linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */


var hasCycle = function(head) {

    if (!head) return false;
    
    let shorter = head;
    let faster = head;
    
    while (faster.next && faster.next.next) {
        shorter = shorter.next;
        faster = faster.next.next;
        if (shorter === faster) return true;
    }
    
    return false;
};










var hasCycleOld = function(head) {

    if (!head) return false;
    
    let cur = head;
    

    while (cur && cur.next) {
        if (cur.visited === true) {
            return true;
        }
        cur.visited = true
        cur = cur.next;
    }
    return false;
    
};

File: 141-linked-list-cycle/141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        '''
        Two pointer - using a slow and a fast pointer
        
        checking if value at each pointer is equal
        
        if not head: return False
        
        slow = head
        fast = head.next
        
        while fast:
            - compare slow.val and fast.val
                - return True if slow.val and fast.val are equal
            - increment slow by 1
            - increment fast by two

        return False
        
        while slow.val != fast.val:
            fast = fast.next.next
            slow = slow.next
            
        time complexity = O(n)
        '''
        
        if not head: return False
        
        slow = head
        fast = head.next
        
        while fast:
            if slow == fast:
                return True
            
            slow = slow.next
            fast = fast.next
            if fast: fast = fast.next
            else: break
            
        return False
        
        